### PR TITLE
Pause shutdown

### PR DESCRIPTION
**Pull request by meejah**
Further to #278 this now allows the controller to wait for subprocesses to shut down, and also handles "GuestWorkerProccess" properly as well which I missed before.

Should I take out the "KILL" code, and just wait forever?

### CODE REVIEW

**Comment by estan** [comment]
This looks like a step in the right direction, but I still get e.g:

```
^C2015-04-01 09:03:51+0200 [Controller   6972] Received SIGINT, shutting down.
2015-04-01 09:03:51+0200 [Controller   6972] sending TERM to subprocess 6975
2015-04-01 09:03:51+0200 [Controller   6972] waiting for 6975 to exit...
2015-04-01 09:03:51+0200 [Controller   6972] sending TERM to subprocess 6980
2015-04-01 09:03:51+0200 [Controller   6972] waiting for 6980 to exit...
2015-04-01 09:03:51+0200 [Controller   6972] sending TERM to subprocess 6981
2015-04-01 09:03:51+0200 [Controller   6972] waiting for 6981 to exit...
2015-04-01 09:03:51+0200 [Container    6981] Received SIGINT, shutting down.
2015-04-01 09:03:51+0200 [Container    6981] Connection to node controller lost.
2015-04-01 09:03:51+0200 [Container    6981] Unhandled Error
2015-04-01 09:03:51+0200 [Container    6981] Traceback (most recent call last):
2015-04-01 09:03:51+0200 [Container    6981] Failure: exceptions.RuntimeError: No transport, but disconnect() called.
2015-04-01 09:03:51+0200 [Container    6981] 'While firing onLeave'
2015-04-01 09:03:51+0200 [Container    6981] Received SIGTERM, shutting down.
2015-04-01 09:03:51+0200 [Container    6981] Unhandled Error
2015-04-01 09:03:51+0200 [Container    6981] Traceback (most recent call last):
2015-04-01 09:03:51+0200 [Container    6981] Failure: exceptions.RuntimeError: No transport, but disconnect() called.
2015-04-01 09:03:51+0200 [Container    6981] 'While firing onLeave'
2015-04-01 09:03:51+0200 [Container    6981] Main loop terminated.
2015-04-01 09:03:51+0200 [Router       6975] Received SIGINT, shutting down.
2015-04-01 09:03:51+0200 [Router       6975] Received SIGTERM, shutting down.
2015-04-01 09:03:51+0200 [Router       6975] Unhandled Error
2015-04-01 09:03:51+0200 [Router       6975] Traceback (most recent call last):
2015-04-01 09:03:51+0200 [Router       6975] File "/home/estan/Projekt/orexplore/dev/Elvis/test_env/lib/python2.7/site-packages/crossbar-0.10.3-py2.7.egg/crossbar/worker/process.py", line 193, in <module>
2015-04-01 09:03:51+0200 [Router       6975] run()
2015-04-01 09:03:51+0200 [Router       6975] File "/home/estan/Projekt/orexplore/dev/Elvis/test_env/lib/python2.7/site-packages/crossbar-0.10.3-py2.7.egg/crossbar/worker/process.py", line 181, in run
2015-04-01 09:03:51+0200 [Router       6975] reactor.run()
2015-04-01 09:03:51+0200 [Router       6975] File "/home/estan/Projekt/orexplore/dev/Elvis/test_env/lib/python2.7/site-packages/twisted/internet/base.py", line 1192, in run
2015-04-01 09:03:51+0200 [Router       6975] self.mainLoop()
2015-04-01 09:03:51+0200 [Router       6975] File "/home/estan/Projekt/orexplore/dev/Elvis/test_env/lib/python2.7/site-packages/twisted/internet/base.py", line 1201, in mainLoop
2015-04-01 09:03:51+0200 [Router       6975] self.runUntilCurrent()
2015-04-01 09:03:51+0200 [Router       6975] --- <exception caught here> ---
2015-04-01 09:03:51+0200 [Router       6975] File "/home/estan/Projekt/orexplore/dev/Elvis/test_env/lib/python2.7/site-packages/twisted/internet/base.py", line 797, in runUntilCurrent
2015-04-01 09:03:51+0200 [Router       6975] f(*a, **kw)
2015-04-01 09:03:51+0200 [Router       6975] File "/home/estan/Projekt/orexplore/dev/Elvis/test_env/lib/python2.7/site-packages/twisted/internet/base.py", line 580, in stop
2015-04-01 09:03:51+0200 [Router       6975] "Can't stop reactor that isn't running.")
2015-04-01 09:03:51+0200 [Router       6975] twisted.internet.error.ReactorNotRunning: Can't stop reactor that isn't running.
2015-04-01 09:03:51+0200 [Router       6975] Connection to node controller lost.
2015-04-01 09:03:51+0200 [Router       6975] Unhandled Error
2015-04-01 09:03:51+0200 [Router       6975] Traceback (most recent call last):
2015-04-01 09:03:51+0200 [Router       6975] Failure: exceptions.RuntimeError: No transport, but disconnect() called.
2015-04-01 09:03:51+0200 [Router       6975] 'While firing onLeave'
2015-04-01 09:03:51+0200 [Router       6975] (TCP Port 8080 Closed)
2015-04-01 09:03:51+0200 [Router       6975] Main loop terminated.
2015-04-01 09:03:51+0200 [Controller   6972] GuestWorkerClientProtocol.connectionLost: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
        ]
2015-04-01 09:03:51+0200 [Controller   6972] GuestWorkerClientProtocol: guest ended cleanly
2015-04-01 09:03:51+0200 [Controller   6972] Guest excited with success
2015-04-01 09:03:51+0200 [Controller   6972] Worker 6981: Process connection gone (Connection was closed cleanly.)
2015-04-01 09:03:51+0200 [Controller   6972] Worker 6975: Process connection gone (Connection was closed cleanly.)
2015-04-01 09:03:51+0200 [Controller   6972] Main loop terminated.
```

when i `Ctrl+C` with the setup I described [here in #278](https://github.com/crossbario/crossbar/issues/278#issuecomment-87978078).

I'm not quite sure what happens, but I suspect the shell is delivering `SIGINT` to the workers, in addition to your `SIGTERM`. I'm not quite sure why this would result in the connection to the node controller to be lost, since by the output, it looks like your code is waiting for them to exit properly.

It's not completely deterministic though. Another time I get:

```
^C2015-04-01 09:10:40+0200 [Controller   7061] Received SIGINT, shutting down.
2015-04-01 09:10:40+0200 [Controller   7061] sending TERM to subprocess 7064
2015-04-01 09:10:40+0200 [Controller   7061] waiting for 7064 to exit...
2015-04-01 09:10:40+0200 [Controller   7061] sending TERM to subprocess 7070
2015-04-01 09:10:40+0200 [Controller   7061] waiting for 7070 to exit...
2015-04-01 09:10:40+0200 [Controller   7061] sending TERM to subprocess 7071
2015-04-01 09:10:40+0200 [Controller   7061] waiting for 7071 to exit...
2015-04-01 09:10:40+0200 [Router       7064] Received SIGINT, shutting down.
2015-04-01 09:10:40+0200 [Router       7064] Received SIGTERM, shutting down.
2015-04-01 09:10:40+0200 [Router       7064] Connection to node controller lost.
2015-04-01 09:10:40+0200 [Container    7071] Received SIGINT, shutting down.
2015-04-01 09:10:40+0200 [Container    7071] Connection to node controller lost.
2015-04-01 09:10:40+0200 [Container    7071] Received SIGTERM, shutting down.
2015-04-01 09:10:40+0200 [Container    7071] Unhandled Error
2015-04-01 09:10:40+0200 [Container    7071] Traceback (most recent call last):
2015-04-01 09:10:40+0200 [Container    7071] Failure: exceptions.RuntimeError: No transport, but disconnect() called.
2015-04-01 09:10:40+0200 [Container    7071] Unhandled Error
2015-04-01 09:10:40+0200 [Container    7071] Traceback (most recent call last):
2015-04-01 09:10:40+0200 [Container    7071] Failure: exceptions.RuntimeError: No transport, but disconnect() called.
2015-04-01 09:10:40+0200 [Container    7071] 'While firing onLeave'
2015-04-01 09:10:40+0200 [Router       7064] Unhandled Error
2015-04-01 09:10:40+0200 [Router       7064] Traceback (most recent call last):
2015-04-01 09:10:40+0200 [Router       7064] Failure: exceptions.RuntimeError: No transport, but disconnect() called.
2015-04-01 09:10:40+0200 [Router       7064] 'While firing onLeave'
2015-04-01 09:10:40+0200 [Router       7064] (TCP Port 8080 Closed)
2015-04-01 09:10:40+0200 [Router       7064] Main loop terminated.
2015-04-01 09:10:40+0200 [Container    7071] Main loop terminated.
2015-04-01 09:10:40+0200 [Controller   7061] GuestWorkerClientProtocol.connectionLost: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ProcessTerminated'>: A process has ended with a probable error condition: process ended by signal 15.
        ]
2015-04-01 09:10:40+0200 [Controller   7061] GuestWorkerClientProtocol: guest ended with error None
2015-04-01 09:10:40+0200 [Controller   7061] Guest excited with error [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ProcessTerminated'>: A process has ended with a probable error condition: process ended by signal 15.
        ]
2015-04-01 09:10:40+0200 [Controller   7061] Worker 7071: Process connection gone (Connection was closed cleanly.)
2015-04-01 09:10:40+0200 [Controller   7061] Worker 7064: Process connection gone (Connection was closed cleanly.)
2015-04-01 09:10:40+0200 [Controller   7061] Main loop terminated.
```

But I guess that's just due to the order in which things happened to happen.

So I still think there's some room to make this more robust. I'm not sure if there's a way to prevent the delivery of `SIGINT` to all child processes (which is standard behavior btw, since I think they are in the foreground process group), and let the controller have full control (no pun intended).

**Comment by oberstet** [comment]
I think we need to test these cases:

1. pressing CTRL-C in the shell where Crossbar.io was started
2. sending a SIGINT to the Crossbar.io controller process (e.g. kill -SIGINT <PID>, where PID is the process id of the controller process)
3. sending a SIGINT to a native worker
4. sending a SIGINT to a guest worker

There is a difference between 1. and 2. - this is explained [here](http://unix.stackexchange.com/a/149756/52500) and [here](http://stackoverflow.com/a/6747510/884770).

Also: I am not sure what signal propagation (to processes spawned from the original Twisted process) is already built into Twisted itself and what we need to manually implement.

**Comment by estan** [comment]
Yes, good list of test cases @oberstet. I'd also do one round through them with `s/SIGINT/SIGTERM/`. Is a router worker just a regular native worker, or is it special in any way? If it is, then maybe better add "5. sending a SIGINT/SIGTERM to a router worker." ?

**Comment by oberstet** [comment]
A router worker is just one of 2 subtypes of native workers. The other being a container worker. There isn't a difference regarding the stuff here. It's sufficient for our needs here if we distinguish these process types:

* controller
* native worker
* guest worker

Signals: yes, we should test SIGINT, SIGTERM and SIGKILL.

**Comment by estan** [comment]
Alright, sounds good. Now I wonder what is happening in my runs above. From the output it seems things happen like this:

1. Controller (since it is the parent process?) receives `SIGINT` first of all.
2. Controller sends `SIGTERM` to workers and starts waiting for them to finish (i.e. @meejah's code is doing what it should).
3. Workers receive `SIGINT`, since they are also part of the foreground process group and starts shutting down.
4. Workers receive the `SIGTERM` sent by the controller and starts shutting down again.

Questions:

1. Why the "Connection to node controller lost." messages? Shouldn't the controller still be up at that point? Or is this a normal shutdown message for workers (e.g. due to the worker closing the connection during shutdown)?
2. I guess the double shutdown in the workers is the reason for the more grave message "No transport, but disconnect() called." and the "Can't stop reactor that isn't running." exception?

**Comment by oberstet** [comment]
@estan Yes, I also supsected that in the CTRL-C scenario we now have some "double shutdown" behavior (which made me Google and come up with above links). All processes receive signals from the shell, because they are in the same foreground group. But probably in arbitrary order. And the workers additionally receive signals because of the explicit code in the controller.

Regarding those transport messages: when the controller starts a worker (both native and guests), there will be pipes between the controller and the worker for stdin/stdout/stderr.

With native workers, those pipes are used for running a WAMP session over. The controller can instruct e.g. a router worker to dynamically ramp up a transport or modify realm etc. This is used for management.

With guest workers, there isn't such management and no WAMP session running over the connected pipes. You can configure what the controller should do with stdin connected to guest worker: close it, send some JSON config and then close, .. and stdout: discard, log, ...

**Comment by oberstet** [comment]
Rgd. controller: yes, this is the parent process of everything. When you type `crossbar start`, this essentially starts the controller process. Depending on config, the controller then starts new processes.

**Comment by estan** [comment]
Ah. I see. But then I think everything bad I see right is due to this double deliverance of SIGINT + SIGTERM some way or another. Should the workers somehow be equipped to handle double termination/interruption, and avoid the double shutdown sequence (I mean, ignore signals if it is already in the process of shutting down)?

**Comment by meejah** [comment]
I will double-check what Twisted does, if anything, for spawned processes.

I do have test-cases for all the cases @oberstet metions, except they're against the router worker (need those repeated with guest workers, too). I also have a guest-worker case that does a "slow" shutdown (i.e. sleep(4) in its signal handler, before exiting). Perhaps I can also make an automated case that simulates the ctrl-c behavior more-exactly. I was testing the guest worker setup both with a "normal" ctrl-c in a terminal, and by sending SIGINT to the controller (directly, not to the process-group).

On my machine, with a setup mirroring @estan's setup, ctrl-C results in an orderly exit. My "worker" just does "sleep(1)" in a forever-loop (and installed its own INT and TERM handlers)...

**Comment by meejah** [comment]
I tried with the "kill" way of delivering a signal to the process-group, as well as "normal" ctrl-C but still see similar behavior. I'll also with a "real" guest worker (i.e. using ApplicationProcess etc).

Twisted just calls ``reactor.stop()`` on INT or TERM, but doesn't do any handling for INT if there's already a custom signal-handler for that installed. So, that's likely the source of the "already stopped" errors. I guess my shell/terminal isn't propagating ctrl-C the same as @estan's.

Perhaps the right behavior is to arrange to ignore SIGINT in the worker processes. not-sending-TERM from the controller works for only the ctrl-C case; sending INT or TERM directly to the controller won't end up killing the child at all (well, until the KILL timeout).

@estan, what OS and Twisted version are you using?

**Comment by meejah** [comment]
I put in some code so native workers ignore SIGINT, although I'm not completely sure if that's the "proper" behavior; will test further.

**Comment by estan** [comment]
Hm. Okay, I have no signal handling in my container worker, but in my guest I do catch KeyboardInterrupt (`SIGINT`) and have a normal signal handler installed for `SIGTERM`.

So you are not seeing neither "No transport, but disconnect() called." nor the "Can't stop reactor that isn't running." exception, @meejah ? And what about the "[Controller   7061] GuestWorkerClientProtocol.connectionLost: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ProcessTerminated'>: A process has ended with a probable error condition: process ended by signal 15."?

I would think the controller logic upon receiving a SIGINT or SIGTERM signal should be (syncronous pseudocode):

```
HANDLE_SIGINT_OR_SIGTERM(signal):
    IF signal == SIGINT AND due to Ctrl+C AND shell is POSIX compliant:
        assume SIGHUP has been delivered to workers and just wait for them
    ELSE:
        propagate signal to workers and wait for them
```

The waiting would of course have a timeout like you have now. (Maybe if workers don't respond to SIGINT, they should be sent SIGTERM, and finally as a last resort SIGKILL? But that's a detail).

**Comment by meejah** [comment]
SIGINT would deliver SIGINT to all the workers, not HUP, right? (but that's a detail and perhaps OS-dependant?). Is there a way for the parent to know if its child has been delivered INT?

I only see the "signal 15" exception if I take out the SIGINT handler in my guest. This is different from just "catch KeyboardInterrupt"; you have to use ``signal`` to install a handler. A do-nothing one would look like:

```python
import signal
signal.signal(signal.SIGINT, lambda sig, frame: None)
```

However, I don't see the "reactor that isn't running" errors, with or without the SIGINT-ignoring code I just added (on a ctrl-C, SIGINT to controller, or SIGINT to process-group).

This is on Debian "jessie"/testing, Twisted 15.0.0, Python 2.7.8

**Comment by estan** [comment]
Yes, sorry, the `SIGHUP` was a typo. And yes, it's very platform (and maybe even shell) specific. I don't know how it works on e.g. Windows, but isn't signals in general a POSIX/*NIX thing?

Ah, yes, I realized after my last message that catching `KeyboardInterrupt` is not the same thing as installing a signal handler for `SIGINT`, so I should do that to avoid the signal 15 message I guess.

Did you try `Ctrl+C` a lot of times? My output is not deterministic, and I only get the "reactor isn't running" sometimes I think.

I'm on Arch Linux with Python 2.7.9, and:

```
(test_env)[estan@pyret Elvis]$ pip list
autobahn (0.10.2)
cffi (0.9.2)
characteristic (14.3.0)
crossbar (0.10.3)
cryptography (0.8.1)
enum34 (1.0.4)
Jinja2 (2.7.3)
MarkupSafe (0.23)
mistune (0.5.1)
netaddr (0.7.13)
pip (6.0.8)
pyasn1 (0.1.7)
pyasn1-modules (0.0.5)
pycparser (2.10)
Pygments (2.0.2)
pyOpenSSL (0.14)
PyTrie (0.2)
PyYAML (3.11)
requests (2.6.0)
service-identity (14.0.0)
setuptools (12.0.5)
shutilwhich (1.0.1)
six (1.9.0)
treq (15.0.0)
Twisted (15.0.0)
zope.interface (4.1.2)
```

(Note the crossbar version isn't actually 0.10.3, but the version in this pull request.)

**Comment by estan** [comment]
I'm not sure if there's a way for the parent to 1) Know if `SIGINT` has been delivered to its children or 2) If the `SIGINT` was due to `Ctrl+C` or e.g. `kill -INT`, or 3) If the current shell is POSIX compliant (which should imply 1), so perhaps my idea above is not really feasible and another way should be found.

**Comment by meejah** [comment]
see also https://github.com/tavendo/AutobahnPython/pull/375

**Comment by estan** [comment]
Yay, with both these pull reqs I can reliably get

```
^C2015-04-03 09:09:17+0200 [Controller  28184] Received SIGINT, shutting down.
2015-04-03 09:09:17+0200 [Controller  28184] sending TERM to subprocess 28187
2015-04-03 09:09:17+0200 [Controller  28184] waiting for 28187 to exit...
2015-04-03 09:09:17+0200 [Controller  28184] sending TERM to subprocess 28192
2015-04-03 09:09:17+0200 [Controller  28184] waiting for 28192 to exit...
2015-04-03 09:09:17+0200 [Controller  28184] sending TERM to subprocess 28193
2015-04-03 09:09:17+0200 [Controller  28184] waiting for 28193 to exit...
2015-04-03 09:09:17+0200 [Container   28193] Ignoring SIGINT in worker.
2015-04-03 09:09:17+0200 [Container   28193] Received SIGTERM, shutting down.
2015-04-03 09:09:17+0200 [Router      28187] Ignoring SIGINT in worker.
2015-04-03 09:09:17+0200 [Router      28187] Received SIGTERM, shutting down.
2015-04-03 09:09:17+0200 [Router      28187] Connection to node controller lost.
2015-04-03 09:09:17+0200 [Router      28187] (TCP Port 8080 Closed)
2015-04-03 09:09:17+0200 [Controller  28184] Guest excited with success
2015-04-03 09:09:17+0200 [Container   28193] Connection to node controller lost.
2015-04-03 09:09:17+0200 [Container   28193] Main loop terminated.
2015-04-03 09:09:17+0200 [Router      28187] Main loop terminated.
2015-04-03 09:09:17+0200 [Controller  28184] Worker 28193: Process connection gone (Connection was closed cleanly.)
2015-04-03 09:09:17+0200 [Controller  28184] Worker 28187: Process connection gone (Connection was closed cleanly.)
2015-04-03 09:09:17+0200 [Controller  28184] Main loop terminated.
```

when hitting `Ctrl+C`.

Great work @meejah.

**Comment by estan** [comment]
For posterity, here's what I reliably get when I `kill -INT` the node controller process:

```
2015-04-03 09:37:46+0200 [Controller  28806] Received SIGINT, shutting down.
2015-04-03 09:37:46+0200 [Controller  28806] sending TERM to subprocess 28809
2015-04-03 09:37:46+0200 [Controller  28806] waiting for 28809 to exit...
2015-04-03 09:37:46+0200 [Controller  28806] sending TERM to subprocess 28814
2015-04-03 09:37:46+0200 [Controller  28806] waiting for 28814 to exit...
2015-04-03 09:37:46+0200 [Controller  28806] sending TERM to subprocess 28815
2015-04-03 09:37:46+0200 [Controller  28806] waiting for 28815 to exit...
2015-04-03 09:37:46+0200 [Router      28809] Received SIGTERM, shutting down.
2015-04-03 09:37:46+0200 [Router      28809] (TCP Port 8080 Closed)
2015-04-03 09:37:46+0200 [Router      28809] Connection to node controller lost.
2015-04-03 09:37:46+0200 [Container   28815] Received SIGTERM, shutting down.
2015-04-03 09:37:46+0200 [Container   28815] Connection to node controller lost.
2015-04-03 09:37:46+0200 [Router      28809] Main loop terminated.
2015-04-03 09:37:46+0200 [Controller  28806] Guest excited with success
2015-04-03 09:37:46+0200 [Container   28815] Main loop terminated.
2015-04-03 09:37:46+0200 [Controller  28806] Worker 28815: Process connection gone (Connection was closed cleanly.)
2015-04-03 09:37:46+0200 [Controller  28806] Worker 28809: Process connection gone (Connection was closed cleanly.)
2015-04-03 09:37:46+0200 [Controller  28806] Main loop terminated.
```

So all is good. Note that container/router workers are no longer "Ignoring SIGINT in worker.".

And here's with `kill -TERM`:

```
2015-04-03 09:42:04+0200 [Controller  28935] Received SIGTERM, shutting down.
2015-04-03 09:42:04+0200 [Controller  28935] sending TERM to subprocess 28938
2015-04-03 09:42:04+0200 [Controller  28935] waiting for 28938 to exit...
2015-04-03 09:42:04+0200 [Controller  28935] sending TERM to subprocess 28943
2015-04-03 09:42:04+0200 [Controller  28935] waiting for 28943 to exit...
2015-04-03 09:42:04+0200 [Controller  28935] sending TERM to subprocess 28944
2015-04-03 09:42:04+0200 [Controller  28935] waiting for 28944 to exit...
2015-04-03 09:42:04+0200 [Router      28938] Received SIGTERM, shutting down.
2015-04-03 09:42:04+0200 [Router      28938] (TCP Port 8080 Closed)
2015-04-03 09:42:04+0200 [Router      28938] Connection to node controller lost.
2015-04-03 09:42:04+0200 [Router      28938] Main loop terminated.
2015-04-03 09:42:04+0200 [Controller  28935] Guest excited with success
2015-04-03 09:42:04+0200 [Container   28944] Received SIGTERM, shutting down.
2015-04-03 09:42:04+0200 [Container   28944] Connection to node controller lost.
2015-04-03 09:42:04+0200 [Container   28944] Main loop terminated.
2015-04-03 09:42:04+0200 [Controller  28935] Worker 28944: Process connection gone (Connection was closed cleanly.)
2015-04-03 09:42:04+0200 [Controller  28935] Worker 28938: Process connection gone (Connection was closed cleanly.)
2015-04-03 09:42:04+0200 [Controller  28935] Main loop terminated.
```

So I'm all happy now.

I also tried with `kill -KILL`, which immediately terminated the controller, but left my guest running. This is to be expected of course, as `SIGKILL` can't be handled by the controller.

File: crossbar/worker/process.py [hunk 1]
```diff
@@ -41,6 +41,20 @@ def run():
...
+    import signal
+
+    # Ignore SIGINT so we get consistent behavior on control-C versus
+    # sending SIGINT to the controller process. When the controller is
+    # shutting down, it sends TERM to all its children but ctrl-C
+    # handling will send a SIGINT to all the processes in the group
+    # (so then the controller sends a TERM but the child already or
+    # will very shortly get a SIGINT as well). Twisted installs SIGINT
+    # handlers, but not for SIGINT if there's already a custom one
+    # present.
```
Comment (estan) [inline crossbar/worker/process.py]: The last sentence in this comment reads a little weird, should it perhaps be "Twisted installs signal handlers, but not for SIGINT if there's already a custom one present." ?

**Comment by oberstet** [comment]
@meejah awesome! seems you've cracked the nut;) could you pls squash to trunk and I merge?

**Comment by meejah** [comment]
See #294

**Pull request closed without merging**